src/passthrough/mod.rs

```diff
@@ -664,3 +888,93 @@ impl<D: AsyncDrive> BackendFileSystem for PassthroughFs<D> {
 fn ebadf() -> io::Error {
     io::Error::from_raw_os_error(libc::EBADF)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::api::filesystem::*;
+    use crate::api::{Vfs, VfsOptions};
+    use log;
+    use std::ops::Deref;
+    use vmm_sys_util::{tempdir::TempDir, tempfile::TempFile};
+
+    fn passthroughfs_no_open(cfg: bool) {
+        let opts = VfsOptions {
+            no_open: cfg,
+            ..Default::default()
+        };
+
+        let vfs = &Vfs::<AsyncDriver, ()>::new(opts);
+        // Assume that fuse kernel supports no_open.
+        vfs.init(FsOptions::ZERO_MESSAGE_OPEN).unwrap();
+
+        let fs_cfg = Config {
+            do_import: false,
+            no_open: cfg,
+            ..Default::default()
+        };
+        let fs = PassthroughFs::new(fs_cfg.clone()).unwrap();
+        fs.import().unwrap();
+        vfs.mount(Box::new(fs), "/submnt/A").unwrap();
+
+        let p_fs = vfs.get_rootfs("/submnt/A").unwrap().unwrap();
+        let any_fs = p_fs.deref().as_any();
+        any_fs
+            .downcast_ref::<PassthroughFs>()
+            .map(|fs| {
+                assert_eq!(fs.no_open.load(Ordering::Relaxed), cfg);
+            })
+            .unwrap();
+    }
+
+    #[test]
+    fn test_passthroughfs_no_open() {
+        passthroughfs_no_open(true);
+        passthroughfs_no_open(false);
+    }
+
+    #[test]
+    fn test_passthroughfs_inode_file_handles() {
+        log::set_max_level(log::LevelFilter::Trace);
+
+        let source = TempDir::new().expect("Cannot create temporary directory.");
+        let parent_path =
+            TempDir::new_in(source.as_path()).expect("Cannot create temporary directory.");
+        let child_path =
+            TempFile::new_in(parent_path.as_path()).expect("Cannot create temporary file.");
+
+        let fs_cfg = Config {
+            writeback: true,
+            do_import: true,
+            no_open: true,
+            inode_file_handles: true,
+            root_dir: source
+                .as_path()
+                .to_str()
+                .expect("source path to string")
+                .to_string(),
+            ..Default::default()
+        };
+        let fs = PassthroughFs::<AsyncDriver, ()>::new(fs_cfg).unwrap();
+        fs.import().unwrap();
+
+        let ctx = Context::default();
+
+        // read a few files to inode map.
+        let parent = CString::new(parent_path.as_path().to_str().expect("path to string")).unwrap();
+        let p_entry = fs.lookup(ctx, ROOT_ID, &parent).unwrap();
+        let p_inode = p_entry.inode;
+
+        let child = CString::new(child_path.as_path().to_str().expect("path to string")).unwrap();
+        let c_entry = fs.lookup(ctx, p_inode, &child).unwrap();
```

Comment:
CI failed here due to "invoking open_by_handle_at needs CAP_DAC_READ_SEARCH".

So I suggest we can add a patch like below,

```
     #[test]
     fn test_passthroughfs_inode_file_handles() {
+        use caps::{CapSet, Capability};
+
+        match caps::has_cap(None, CapSet::Effective, Capability::CAP_DAC_READ_SEARCH) {
+            Ok(true) => {}
+            Ok(false) | Err(_) => {
+                println!("invoking open_by_handle_at needs CAP_DAC_READ_SEARCH");
+                return;
+            }
+        }
+
```